### PR TITLE
Composable requirements.

### PR DESCRIPTION
**Pull request by DorianGray**
Since we have a varied set of possible dependencies based on features
used, it probably makes sense to break up our requirements into
composable files. This also unifies our requirements into exactly one
place.

### CODE REVIEW

**Comment by DorianGray** [comment]
Please let me know if I've missed any (optional) dependencies. It would be nice to have them all documented at least for pip.

File: requirements/base [hunk 1]
```diff
@@ -0,0 +1,4 @@
+-r xcffib
```
Comment (ramnes) [inline requirements/base:1]: Is this a working hack so that xcffib is always installed first?!

File: requirements/doc [hunk 2]
```diff
@@ -0,0 +1,6 @@
+-r base
+setuptools_scm
+sphinx
+sphinx_rtd_theme
+sphinxcontrib-seqdiag
+numpydoc
```
Comment (ramnes) [inline requirements/doc:6]: Let's call this file "docs" to stay consistent

File: requirements/base (hunk 1 above)
Comment (DorianGray) [reply requirements/base:1]: yeah

File: requirements/lint [hunk 3]
```diff
@@ -0,0 +1,4 @@
+flake8
+flake8-isort
```
Comment (ramnes) [inline requirements/lint:2]: I think `isort` will have give troubles without a `-r base` here

File: requirements/static [hunk 4]
```diff
@@ -0,0 +1,3 @@
+-r base
+-r optional
+mypy
```
Comment (ramnes) [inline requirements/static:3]: Why not put mypy in lint as well?

File: requirements/test [hunk 5]
```diff
@@ -0,0 +1,4 @@
+-r base
+pytest
+pytest-cov
+pytest-runner
```
Comment (ramnes) [inline requirements/test:4]: the PR to remove pytest-runner has been merged

File: tox.ini [hunk 6]
```diff
@@ -16,55 +16,41 @@ envlist =
 # that our tests use.
 setenv = LC_CTYPE = en_US.UTF-8
 # Pass Display down to have it for the tests available
-passenv = DISPLAY TRAVIS TRAVIS_JOB_ID TRAVIS_BRANCH
```
Comment (ramnes) [inline tox.ini:19]: good catch

File: tox.ini [hunk 7]
```diff
@@ -16,55 +16,41 @@ envlist =
...
-    pytest >= 6.2.1
-    pytest-cov >= 2.10.1
-    setuptools >= 40.5.0
-    xcffib >= 0.8.1
+passenv = DISPLAY
+# xcffib has to be installed before cairocffi, so we can't use deps anywhere else
+deps = -r requirements/xcffib
 commands =
-    python3 setup.py install
+    pip install -q -r requirements/test -e .
```
Comment (ramnes) [inline tox.ini:23]: why -q and -e?

File: requirements/static (hunk 4 above)
Comment (DorianGray) [reply requirements/static:3]: lint and static analysis are separate jobs right now, should I combine them? I wanted these files to be composable so if we decide to parallelize more it's easy

Comment (ramnes) [reply requirements/static:3]: Forget it

File: requirements/test (hunk 5 above)
Comment (DorianGray) [reply requirements/test:4]: ty, will fix....I thought I resolved all conflicts but missed one!

File: .github/workflows/ci.yml [hunk 8]
```diff
@@ -21,7 +21,7 @@ jobs:
                   python-version: ${{ matrix.python-version }}
             - run: |
                 sudo apt update
-                sudo apt install graphviz imagemagick x11-apps xserver-xephyr xterm xvfb libpulse-dev lm-sensors
+                sudo apt install libdbus-1-dev graphviz imagemagick x11-apps xserver-xephyr xterm xvfb libpulse-dev lm-sensors
```
Comment (ramnes) [inline .github/workflows/ci.yml:24]: Why this change?

File: MANIFEST.in [hunk 9]
```diff
@@ -9,6 +9,7 @@ exclude .pylintrc
 exclude tox.ini
 exclude requirements.txt
 exclude requirements-dev.txt
+exclude requirements/*
```
Comment (ramnes) [inline MANIFEST.in:12]:
```suggestion
exclude requirements/
```

File: .github/workflows/ci.yml (hunk 8 above)
Comment (DorianGray) [reply .github/workflows/ci.yml:24]: we weren't statically analyzing dbus stuff before because missing library, now we are!

File: MANIFEST.in (hunk 9 above)
Comment (DorianGray) [reply MANIFEST.in:12]: trailing backslashes in this doesn't work on windows and causes a warning. windows doesn't matter, but the warning does.

File: docs/requirements.txt [hunk 10]
```diff
@@ -1,7 +1 @@
-setuptools_scm
-sphinx
-sphinx_rtd_theme
-sphinxcontrib-seqdiag
-numpydoc
-
--r ../requirements.txt
+-r ../requirements/doc
```
Comment (ramnes) [inline docs/requirements.txt:1]: I'd rather try to remove this file completely, or symlink

File: tox.ini (hunk 7 above)
Comment (DorianGray) [reply tox.ini:23]: -q hides the output, pip output seems unnecessarily loud for our purposes. the `-e .` is instead of calling `python3 setup.py install`

File: requirements.txt [hunk 11]
```diff
@@ -1,3 +1 @@
-cairocffi >= 0.9.0
-cffi >= 1.1.0
-xcffib >= 0.10.1
+-r requirements/base
```
Comment (ramnes) [inline requirements.txt:1]: Let's remove this file

File: requirements-dev.txt [hunk 12]
```diff
@@ -1,7 +1,3 @@
-flake8
-flake8-isort
-flake8-tidy-imports
-mypy
-pep8-naming
-psutil
-pytest-cov
+-r requirements/base
+-r requirements/test
+-r requirements/static
```
Comment (ramnes) [inline requirements-dev.txt:3]: Let's remove this file

File: tox.ini (hunk 7 above)
Comment (ramnes) [reply tox.ini:23]: you can `pip install .` without `-e`, and I don't think the editable mode is necessary here

File: .github/workflows/ci.yml (hunk 8 above)
Comment (ramnes) [reply .github/workflows/ci.yml:24]: :tada: 

please make it a separate commit with that message so that we can remember why we have this

File: MANIFEST.in (hunk 9 above)
Comment (ramnes) [reply MANIFEST.in:12]: oh really? then sorry for the noise :)

File: requirements/lint (hunk 3 above)
Comment (DorianGray) [reply requirements/lint:2]: troubles how so? It works in practice.

File: requirements/base (hunk 1 above)
Comment (DorianGray) [reply requirements/base:1]: err, no, sorry, the hack is in tox.ini

Comment (DorianGray) [reply requirements/base:1]: this is just for completeness, an in hopes someday it will just work

File: tox.ini (hunk 7 above)
Comment (DorianGray) [reply tox.ini:23]: fair enough

File: .github/workflows/ci.yml (hunk 8 above)
Comment (DorianGray) [reply .github/workflows/ci.yml:24]: :+1:

File: requirements/base (hunk 1 above)
Comment (ramnes) [reply requirements/base:1]: my hopes were high but alas no, that doesn't work:

```
$ echo xcffib > xcffib
$ echo -e "-r xcffib\ncairocffi\n" > hack
$ pip install --no-cache-dir -r hack
Collecting cairocffi
  Downloading cairocffi-1.2.0.tar.gz (70 kB)
     |████████████████████████████████| 70 kB 1.4 MB/s 
Collecting cffi>=1.1.0
  Downloading cffi-1.14.4-cp37-cp37m-manylinux1_x86_64.whl (402 kB)
     |████████████████████████████████| 402 kB 2.2 MB/s 
Collecting xcffib
  Downloading xcffib-0.11.1.tar.gz (84 kB)
     |████████████████████████████████| 84 kB 3.7 MB/s 
Collecting pycparser
  Downloading pycparser-2.20-py2.py3-none-any.whl (112 kB)
     |████████████████████████████████| 112 kB 3.9 MB/s 
Collecting six
  Downloading six-1.15.0-py2.py3-none-any.whl (10 kB)
Building wheels for collected packages: cairocffi, xcffib
  Building wheel for cairocffi (setup.py) ... done
  Created wheel for cairocffi: filename=cairocffi-1.2.0-py3-none-any.whl size=89545 sha256=ff339b46df2ebf5df7355fef6763c1a6aaf8b9daf904c026d970125e2ec7a278
  Stored in directory: /tmp/pip-ephem-wheel-cache-uk3riam1/wheels/e2/ca/86/9db2824f203afe4bdf5aa6ead017f352fcc90e92f095b78871
  Building wheel for xcffib (setup.py) ... done
  Created wheel for xcffib: filename=xcffib-0.11.1-py3-none-any.whl size=95223 sha256=cadfd8f4d75dcae8b211f81f603f03c12b0e97409b0632fa3bbca6e8d30c6178
  Stored in directory: /tmp/pip-ephem-wheel-cache-uk3riam1/wheels/c6/55/fd/d2964bac524a60cb68d24b09d4cc845c47cb4e4e21f852c10a
Successfully built cairocffi xcffib
Installing collected packages: pycparser, six, cffi, xcffib, cairocffi
Successfully installed cairocffi-1.2.0 cffi-1.14.4 pycparser-2.20 six-1.15.0 xcffib-0.11.1
```

Comment (DorianGray) [reply requirements/base:1]: sorry I was confused, I added comments above.

File: requirements/lint (hunk 3 above)
Comment (ramnes) [reply requirements/lint:2]: Yes, I just tried it and I can confirm. It used to classify as first or third party differently depending on whether the library was available or not in the current environment, but it seems that it's no longer the case.

File: requirements/base (hunk 1 above)
Comment (ramnes) [reply requirements/base:1]: then maybe we should just add xcffib in base and remove that file? I don't see a need for a separate file

**Comment by ramnes** [comment]
`dev.sh` and the docs need an update. And I think I'd prefer to keep the `.txt` suffix on files for the explicitness and because it seems quite standard.

File: requirements/base (hunk 1 above)
Comment (DorianGray) [reply requirements/base:1]: the only reason is to have the version requirement for xcffib in one place instead of two

**Comment by DorianGray** [comment]
Done! .txt has always annoyed me as entirely unnecessary but you're right :/

**Review by ramnes — APPROVED** [review]
Cooool

**Comment by ramnes** [comment]
@tych0 thoughts?

**Comment by tych0** [comment]
I don't have any strong opinions on this one (although I think maybe it needs a rebase -- don't we already have the extra docs building PR?).

**Comment by DorianGray** [comment]
rebased.

**Comment by ramnes** [comment]
Keeping this one aside for after the incoming release.
